File: python/70.py
# Basic DP problem.
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 1:
            return 1
        methods = [0] * n
        methods[0] = 1
        methods[1] = 2
        for i in range(2, n):
            methods[i] = methods[i - 1] + methods[i - 2]
        return methods[n - 1]


# Fibonacci number, Closed-form expression
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        sqrt5 = math.sqrt(5)
        return int((((1 + sqrt5) / 2) ** (n + 1) - ((1 - sqrt5) / 2) ** (n + 1)) / sqrt5)

File: solution/go/70.go
/*
Fibonacci
*/
func climbStairs(n int) int {
    stats := [3]int{1, 2, 3}
    if n <= 3 {
        return stats[n-1]
    }
    for i := 4; i <= n; i++ {
        temp := stats[1] + stats[2]
        stats[0] = stats[1]
        stats[1] = stats[2]
        stats[2] = temp
    }
    return stats[2]
}
